api/src/main/java/org/apache/iceberg/expressions/BoundSetPredicate.java

```diff
@@ -0,0 +1,176 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.expressions;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Set;
+import org.apache.iceberg.util.CharSequenceWrapper;
+
+public class BoundSetPredicate<T> extends Predicate<BoundReference<T>> {
+  private final LiteralSet<T> literalSet;
+
+  BoundSetPredicate(Operation op, BoundReference<T> ref, Set<Literal<T>> lits) {
+    super(op, ref);
+    Preconditions.checkArgument(op == Operation.IN || op == Operation.NOT_IN,
+        "%s predicate does not support a set of literals", op);
+    this.literalSet = new LiteralSet<>(lits);
+  }
+
+  BoundSetPredicate(Operation op, BoundReference<T> ref, LiteralSet<T> lits) {
+    super(op, ref);
+    Preconditions.checkArgument(op == Operation.IN || op == Operation.NOT_IN,
+        "%s predicate does not support a literal set", op);
+    this.literalSet = lits;
+  }
+
+  @Override
+  public Expression negate() {
+    return new BoundSetPredicate<>(op().negate(), ref(), literalSet);
+  }
+
+  public Set<T> literalSet() {
+    return literalSet;
+  }
+
+  @Override
+  String literalString() {
+    return literalSet.toString();
+  }
+
+  /**
+   * Represents a set of literal values in an IN or NOT_IN predicate
+   * @param <T> The Java type of the value, which can be wrapped by a {@link Literal}
+   */
+  private static class LiteralSet<T> implements Set<T>, Serializable {
+    private final Set<T> values;
+
+    @SuppressWarnings("unchecked")
+    LiteralSet(Set<Literal<T>> lits) {
+      Preconditions.checkArgument(lits == null || lits.size() > 1,
+          "The input literal set must include more than 1 element.");
+      values = ImmutableSet.<T>builder().addAll(
+          lits.stream().map(
+              lit -> {
+                if (lit instanceof Literals.StringLiteral) {
+                  return (T) CharSequenceWrapper.wrap(((Literals.StringLiteral) lit).value());
+                } else {
+                  return lit.value();
+                }
+              }
+          ).iterator()).build();
+    }
+
+    @Override
+    public String toString() {
+      return Joiner.on(", ").join(values);
+    }
+
+    @Override
+    public boolean contains(Object object) {
```

Comment:
`CharSeqLiteralSet` should override this. If the object isn't a `CharSequence` then it should return false. If it is, then it should use a thread-local wrapper:

```java
  private static final ThreadLocal<CharSequenceWrapper> wrapper =
      ThreadLocal.withInitial(() -> CharSequenceWrapper.wrap(null));

  @Override
  public boolean contains(Object object) {
    if (object instanceof CharSequence) {
      return values.contains(wrapper.get().set((CharSequence) object));
    }
    return false;
  }
```